Create npm2deb debian bullseye branch

naruxde opened the pull request:
Please take a look at this. The Debian files were created by npm2deb and received a little polish.

Please also test the package that you can build from this branch or find in our OneDrive!

What I find very nice is the line break and the formatting in the control file at e.g. Build-Depends @tboehler1 @iluminat23 - I am almost inclined to propose this as a new standard in our projects 🤣

The changelog contains the complete Git history - It's a first suggestion.

naruxde committed:
packaging: Add npm2deb debian files

These Debian files were generated via npm2deb. The program has been
written for the creation of Debian packages from npm packages.

Signed-off-by: Sven Sager <s.sager@kunbus.com>

tboehler1 commented on debian/upstream/metadata:
```diff
@@ -0,0 +1,7 @@
+---
+Archive: GitHub
+Bug-Database: https://github.com/erminas/node-red-contrib-revpi-nodes/issues
+Contact: https://github.com/erminas/node-red-contrib-revpi-nodes/issues
+Name: node-red-contrib-revpi-nodes
+Repository: https://github.com/erminas/node-red-contrib-revpi-nodes.git
+Repository-Browse: https://github.com/erminas/node-red-contrib-revpi-nodes
```
Is upstream (e.g. erminas) the correct contact for issues and should the links in `Repository` and `Repository-Browse` lead to them, not to us? If I'm understanding correctly the repository served is this one (RevolutionPi/node-red-...), not the one from erminas.

iluminat23 replied:
No, this is correct as it is the upstream metadata not the package metadata.

tboehler1 commented on debian/changelog:
```diff
@@ -0,0 +1,70 @@
+node-red-contrib-revpi-nodes (1.1.0-1+revpi11+1) stable; urgency=medium
```
```suggestion
node-red-contrib-revpi-nodes (1.1.0-1+revpi11+1) bullseye; urgency=medium
```

tboehler1 commented on debian/control:
```diff
@@ -0,0 +1,28 @@
+Priority: optional
+Maintainer: KUNBUS GmbH <support@kunbus.com>
+Rules-Requires-Root: no
+Homepage: https://github.com/erminas/node-red-contrib-revpi-nodes
+Vcs-Browser: https://github.com/erminas/node-red-contrib-revpi-nodes
+Vcs-Git: https://github.com/erminas/node-red-contrib-revpi-nodes
```
Is this the correct information, or should this repository be given instead?

iluminat23 replied:
```suggestion
Vcs-Git: https://github.com/erminas/node-red-contrib-revpi-nodes.git
```

tboehler1 commented on debian/copyright:
```diff
@@ -0,0 +1,46 @@
+Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
+Upstream-Name: node-red-contrib-revpi-nodes
+Upstream-Contact: https://github.com/erminas/node-red-contrib-revpi-nodes/issues
+Source: https://github.com/erminas/node-red-contrib-revpi-nodes
```
Same here, is this the correct information?

naruxde committed:
Import depending node module ws version 8.8.1

The revpi-nodes depend on a specific version of the ws module. This is
imported into our upstream branch to make this dependency available in
packaging with our module. A fixed import is selected to also track
changes in dependencies.

Signed-off-by: Sven Sager <s.sager@kunbus.com>

tboehler1 reviewed (changes requested):
All in all I'm pleasantly surprised with the result of `npm2deb`, there are just a few things I'm unsure of.

Regarding the format of the `Build-Depends` and `Depends` fields: I don't mind staying with the old format and I don't mind the "new" one, but for the sake of consistency I think we should probably stick with the "old" format where the comma is at the end of a line.

iluminat23 commented:
I find the `,` at the begin of the line more disturbing. As it reminds me of the semicolon (`;`) comments in some config files. And I need to remind my self that this is not the case here.

naruxde committed:
packaging: Use the upstream tags in gbp.conf

We have our own changes in the upstream branch, for which we no longer
use the version tags of the provider, but our own.

Signed-off-by: Sven Sager <s.sager@kunbus.com>

naruxde committed:
packaging: Move ws module to local node_modules dir

The integrated websocket module (ws) must be set up as a local
dependency for our nodes. For this purpose, it is moved to the
`node_modules` folder after creation by nodejs (--with nodejs) within
our module.

Signed-off-by: Sven Sager <s.sager@kunbus.com>

naruxde committed:
Create changelog for 1.1.0+ds1-1+revpi11+1 release

Signed-off-by: Sven Sager <s.sager@kunbus.com>

tboehler1 commented on debian/rules:
```diff
@@ -0,0 +1,18 @@
+
+override_dh_auto_install:
+	dh_auto_install --buildsystem=nodejs
+	mkdir -p $(DEB_DESTDIR)/usr/lib/node_modules
+	mv $(DEB_DESTDIR)/usr/share/nodejs/node-red-contrib-revpi-nodes \
+	   $(DEB_DESTDIR)/usr/lib/node_modules
+	mkdir -p $(DEB_DESTDIR)/usr/lib/node_modules/node-red-contrib-revpi-nodes/node_modules
+	mv $(DEB_DESTDIR)/usr/lib/node_modules/node-red-contrib-revpi-nodes/ws \
+	   $(DEB_DESTDIR)/usr/lib/node_modules/node-red-contrib-revpi-nodes/node_modules
+	rmdir $(DEB_DESTDIR)/usr/share/nodejs
```
`/usr/lib/node_modules` seems to be the correct path for npm's global `node_modules` directory on the RevPi. The path can be retrieved with `npm root -g`, and this shows a different prefix (`/usr/local/lib/node_modules`) for my laptop.

**Assuming** that Node-RED looks for additional nodes in npm's global `node_modules` directory this path can't be hard coded like it is here and needs to be adjusted. Additionally, Debian specifies directories where node modules should live here: https://wiki.debian.org/Javascript/Nodejs#Packaging_modules
While testing these directories Node-RED didn't pick up the nodes, though.

tboehler1 reviewed (approved):
I wasn't able to find documentation for Node-RED on where it looks for nodes exactly, so testing manually is all I had. The only working method for installing nodes was in the global npm root (`npm root -g`).
Do keep in mind, though, that the global npm root is not the directory that Debian usually installs modules.

Not a blocker, just more information. The directories are fine by me, but do wait for the review from @iluminat23.
